### PR TITLE
[AOS 705] new motion validator for approaches

### PR DESCRIPTION
**Pull request by radu-omni**

### CODE REVIEW

File: src/ompl/base/src/DiscreteMotionValidator.cpp [hunk 1]
```diff
@@ -48,6 +48,8 @@ void ompl::base::DiscreteMotionValidator::defaultSettings()
 bool ompl::base::DiscreteMotionValidator::checkMotion(const State *s1, const State *s2,
                                                       std::pair<State *, double> &lastValid) const
 {
+    OMPL_DEBUG("PROBLEM!!! CALLING check motion in OMPL!");
```
Comment (SepehrGharaaty) [inline src/ompl/base/src/DiscreteMotionValidator.cpp:51]: remove

File: src/ompl/base/src/DiscreteMotionValidator.cpp [hunk 2]
```diff
@@ -92,6 +94,7 @@ bool ompl::base::DiscreteMotionValidator::checkMotion(const State *s1, const Sta
 
 bool ompl::base::DiscreteMotionValidator::checkMotion(const State *s1, const State *s2) const
 {
+    OMPL_DEBUG("PROBLEM!!! CALLING check motion in OMPL!");
```
Comment (SepehrGharaaty) [inline src/ompl/base/src/DiscreteMotionValidator.cpp:97]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 3]
```diff
@@ -79,6 +79,9 @@ void ompl::geometric::PathSimplifier::freeStates(bool flag)
 /* Based on COMP450 2010 project of Yun Yu and Linda Hill (Rice University) */
 void ompl::geometric::PathSimplifier::smoothBSpline(PathGeometric &path, unsigned int maxSteps, double minChange)
 {
+    auto start = std::chrono::steady_clock::now();
+    std::string msg = "Start SmoothBSpline with: " + std::to_string(path.getStateCount()) + " states and path_length " + std::to_string(path.length());
+    OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:84]: comment out these

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 4]
```diff
@@ -90,9 +93,11 @@ void ompl::geometric::PathSimplifier::smoothBSpline(PathGeometric &path, unsigne
 
     for (unsigned int s = 0; s < maxSteps; ++s)
     {
+        std::cout << "smoothBSpline: starting step : " + std::to_string(s) << std::endl;
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:96]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 5]
```diff
@@ -90,9 +93,11 @@ void ompl::geometric::PathSimplifier::smoothBSpline(PathGeometric &path, unsigne
...
         path.subdivide();
 
         unsigned int i = 2, u = 0, n1 = states.size() - 1;
+        std::cout << "smoothBSpline: n1 is : " + std::to_string(n1) << std::endl;
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:100]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 6]
```diff
@@ -104,27 +109,45 @@ void ompl::geometric::PathSimplifier::smoothBSpline(PathGeometric &path, unsigne
                 {
                     if (si->distance(states[i], temp1) > minChange)
                     {
+
+                        std::cout << "smoothBSpline: distance " + std::to_string(si->distance(states[i], temp1)) + " is greater than min change: " + std::to_string(minChange) << std::endl;
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:113]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 7]
```diff
@@ -104,27 +109,45 @@ void ompl::geometric::PathSimplifier::smoothBSpline(PathGeometric &path, unsigne
...
                         si->copyState(states[i], temp1);
                         ++u;
                     }
+                    else
+                    {
+                        std::cout << "smoothBSpline: NOT GOOD! distance " + std::to_string(si->distance(states[i], temp1)) +
+                                         " is less than min change: " + std::to_string(minChange)
+                                  << std::endl;
+                    }
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:122]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 8]
```diff
@@ -104,27 +109,45 @@ void ompl::geometric::PathSimplifier::smoothBSpline(PathGeometric &path, unsigne
...
                 }
             }
 
             i += 2;
         }
 
         if (u == 0)
+        {
+            std::cout << "smoothBSpline: BREAK is called at step : " + std::to_string(s) << std::endl;
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:131]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 9]
```diff
@@ -104,27 +109,45 @@ void ompl::geometric::PathSimplifier::smoothBSpline(PathGeometric &path, unsigne
...
             break;
+        }
     }
 
     si->freeState(temp1);
     si->freeState(temp2);
+    auto end = std::chrono::steady_clock::now();
+    auto elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
+    std::string end_msg = "END of SmoothBSpline! Time elapsed: " + std::to_string(elapsed_time) + " millisecond, nb of states: " + std::to_string(path.getStateCount()) + " and path_length " + std::to_string(path.length());
+    OMPL_DEBUG(end_msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:141]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 10]
```diff
@@ -104,27 +109,45 @@ void ompl::geometric::PathSimplifier::smoothBSpline(PathGeometric &path, unsigne
...
 }
 
 bool ompl::geometric::PathSimplifier::reduceVertices(PathGeometric &path, unsigned int maxSteps,
                                                      unsigned int maxEmptySteps, double rangeRatio)
 {
-    if (path.getStateCount() < 3)
+    auto start = std::chrono::steady_clock::now();
+    std::string msg = "Start ReduceVertices with: " + std::to_string(path.getStateCount()) + " states and path_length " + std::to_string(path.length());
+    OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:149]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 11]
```diff
@@ -181,13 +209,21 @@ bool ompl::geometric::PathSimplifier::reduceVertices(PathGeometric &path, unsign
                 result = true;
             }
         }
+
+    auto end = std::chrono::steady_clock::now();
+    auto elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
+    std::string end_msg = "END of ReduceVertices! Time elapsed: " + std::to_string(elapsed_time) + " millisecond, nb of states: " + std::to_string(path.getStateCount()) + " and path_length " + std::to_string(path.length());
+    OMPL_DEBUG(end_msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:216]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 12]
```diff
@@ -181,13 +209,21 @@ bool ompl::geometric::PathSimplifier::reduceVertices(PathGeometric &path, unsign
...
     return result;
 }
 
 bool ompl::geometric::PathSimplifier::shortcutPath(PathGeometric &path, unsigned int maxSteps,
                                                    unsigned int maxEmptySteps, double rangeRatio, double snapToVertex)
 {
-    if (path.getStateCount() < 3)
+    auto start = std::chrono::steady_clock::now();
+    std::string msg = "Start ShortcutPath with: " + std::to_string(path.getStateCount()) + " states and path_length " + std::to_string(path.length());
+    OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:225]: remove/comment out

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 13]
```diff
@@ -213,6 +249,9 @@ bool ompl::geometric::PathSimplifier::shortcutPath(PathGeometric &path, unsigned
     // The range (distance) of a single connection that will be attempted
     double rd = rangeRatio * dists.back();
 
+    std::string msg2 = "ShortcutPath: Threshold is" + std::to_string(threshold) + " And rd is " + std::to_string(rd);
+    OMPL_DEBUG(msg2.c_str());
+
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:254]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 14]
```diff
@@ -380,6 +419,11 @@ bool ompl::geometric::PathSimplifier::shortcutPath(PathGeometric &path, unsigned
 
     si->freeState(temp1);
     si->freeState(temp0);
+
+    auto end = std::chrono::steady_clock::now();
+    auto elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
+    std::string end_msg = "END of ShortcutPath! Time elapsed: " + std::to_string(elapsed_time) + " millisecond, nb of states: " + std::to_string(path.getStateCount()) + " and path_length " + std::to_string(path.length());
+    OMPL_DEBUG(end_msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:426]: comment out

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 15]
```diff
@@ -605,7 +650,10 @@ bool ompl::geometric::PathSimplifier::perturbPath(PathGeometric &path, double st
 bool ompl::geometric::PathSimplifier::collapseCloseVertices(PathGeometric &path, unsigned int maxSteps,
                                                             unsigned int maxEmptySteps)
 {
-    if (path.getStateCount() < 3)
+    auto start = std::chrono::steady_clock::now();
+    std::string msg = "Start collapseCloseVertices with: " + std::to_string(path.getStateCount()) + " states and path_length " + std::to_string(path.length());
+    OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:655]: comment out

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 16]
```diff
@@ -660,6 +708,11 @@ bool ompl::geometric::PathSimplifier::collapseCloseVertices(PathGeometric &path,
         else
             break;
     }
+
+    auto end = std::chrono::steady_clock::now();
+    auto elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
+    std::string end_msg = "END of collapseCloseVertices! Time elapsed: " + std::to_string(elapsed_time) + " millisecond, nb of states: " + std::to_string(path.getStateCount()) + " and path_length " + std::to_string(path.length());
+    OMPL_DEBUG(end_msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:715]: comment out

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 17]
```diff
@@ -690,19 +743,39 @@ bool ompl::geometric::PathSimplifier::simplify(PathGeometric &path, const base::
             unsigned int times = 0;
             do
             {
+                std::string msg = "Call shortcutPath " + std::to_string(times) +
+                                  " for path with: " + std::to_string(path.getStateCount()) +
+                                  " states and path_length " + std::to_string(path.length());
+                OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:749]: comment out

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 18]
```diff
@@ -690,19 +743,39 @@ bool ompl::geometric::PathSimplifier::simplify(PathGeometric &path, const base::
...
                 bool shortcut = shortcutPath(path);  // split path segments, not just vertices
                 bool better_goal =
                     gsr_ ? findBetterGoal(path, ptc) : false;  // Try to connect the path to a closer goal
 
                 metricTryMore = shortcut || better_goal;
             } while ((ptc == false || atLeastOnce) && ++times <= 5 && metricTryMore);
 
+            std::string msg = "AFTER shortcutPath : path with: " + std::to_string(path.getStateCount()) +
+                              " states and path_length " + std::to_string(path.length());
+            OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:759]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 19]
```diff
@@ -690,19 +743,39 @@ bool ompl::geometric::PathSimplifier::simplify(PathGeometric &path, const base::
...
+
+
             // smooth the path with BSpline interpolation
             if (ptc == false || atLeastOnce)
+            {
+                auto path_length = path.length();
+                auto state_count = path.getStateCount();
+                std::string msg = "Call SmoothBSpline for path with: " + std::to_string(state_count) +
+                                  " states and path_length " + std::to_string(path_length);
+                OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:769]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 20]
```diff
@@ -690,19 +743,39 @@ bool ompl::geometric::PathSimplifier::simplify(PathGeometric &path, const base::
...
                 smoothBSpline(path, 3, path.length() / 100.0);
+            }
 
             if (ptc == false || atLeastOnce)
             {
+                std::string msg = "Call checkAndRepair for path with: " + std::to_string(path.getStateCount()) +
+                                  " states and path_length " + std::to_string(path.length());
+                OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:777]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 21]
```diff
@@ -719,19 +792,38 @@ bool ompl::geometric::PathSimplifier::simplify(PathGeometric &path, const base::
...
+            OMPL_DEBUG(msg.c_str());
             tryMore = reduceVertices(path);
 
+            }
+
         if ((ptc == false || atLeastOnce) && si_->getStateSpace()->isMetricSpace())
         {
+            std::string msg = "Space is METRIC! Call checkAndRepair for path with: " + std::to_string(path.getStateCount()) +
+                              " states and path_length " + std::to_string(path.length());
+            OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:826]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 22]
```diff
@@ -719,19 +792,38 @@ bool ompl::geometric::PathSimplifier::simplify(PathGeometric &path, const base::
...
             collapseCloseVertices(path);
+            }
 
         // try to reduce verices some more, if there is any point in doing so
         unsigned int times = 0;
         while ((ptc == false || atLeastOnce) && tryMore && ++times <= 5)
+        {
+            std::string msg = "Call reduceVertices in while loop, repeat nb: " +std::to_string(times)+"  for path with: " + std::to_string(path.getStateCount()) +
+                              " states and path_length " + std::to_string(path.length());
+            OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:817]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 23]
```diff
@@ -719,19 +792,38 @@ bool ompl::geometric::PathSimplifier::simplify(PathGeometric &path, const base::
...
             tryMore = reduceVertices(path);
+        }
 
         // try to collapse close-by vertices
         if (ptc == false || atLeastOnce)
+        {
+            std::string msg = "Call collapseCloseVertices for path with: " + std::to_string(path.getStateCount()) +
+                              " states and path_length " + std::to_string(path.length());
+            OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:807]: remove

File: src/ompl/geometric/src/PathSimplifier.cpp [hunk 24]
```diff
@@ -719,19 +792,38 @@ bool ompl::geometric::PathSimplifier::simplify(PathGeometric &path, const base::
 
         // try a randomized step of connecting vertices
         if (ptc == false || atLeastOnce)
+        {
+            std::string msg = "Call reduceVertices for path with: " + std::to_string(path.getStateCount()) +
+                              " states and path_length " + std::to_string(path.length());
+            OMPL_DEBUG(msg.c_str());
```
Comment (SepehrGharaaty) [inline src/ompl/geometric/src/PathSimplifier.cpp:798]: remove

**Review by SepehrGharaaty — CHANGES_REQUESTED** [review]

**Comment by PierreLucGriffin** [comment]
More tests need to be done. On hold until then.

**Pull request closed without merging**